### PR TITLE
Send contacts sync messages whenever the contacts change.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/contact/OWSContactsManager.m [hunk 1]
```diff
@@ -587,6 +587,11 @@ - (UIImage * _Nullable)imageForPhoneIdentifier:(NSString * _Nullable)identifier
     return contact.image;
 }
 
+- (BOOL)hasAddressBook
+{
+    return self.addressBookReference;
```
Comment (michaelkirk) [inline Signal/src/contact/OWSContactsManager.m]: nit: cast this to bool?

File: Signal/src/util/OWSContactsSyncing.m [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+#import "OWSContactsSyncing.h"
+#import "OWSContactsManager.h"
+#import "TSAccountManager.h"
+#import <SignalServiceKit/MIMETypeUtil.h>
+#import <SignalServiceKit/OWSMessageSender.h>
+#import <SignalServiceKit/OWSSyncContactsMessage.h>
+#import <SignalServiceKit/TSAccountManager.h>
+#import <SignalServiceKit/TSStorageManager.h>
+
```
Comment (michaelkirk) [inline Signal/src/util/OWSContactsSyncing.m]: nullability annotations

File: Signal/src/util/OWSContactsSyncing.m [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+                                             untilCancelled:nil];
+
+    return self;
+}
+
+#pragma mark - Methods
+
+- (void)sendSyncContactsMessageIfNecessary
+{
+    if (self.isRequestInFlight) {
```
Comment (michaelkirk) [inline Signal/src/util/OWSContactsSyncing.m]: nit: Can we AssertIsOnMainThread()

**Comment by charlesmchen** [comment]
@michaelkirk Crap - I should have written this in Swift.

File: Signal/src/contact/OWSContactsManager.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/contact/OWSContactsManager.m]: Done.

File: Signal/src/util/OWSContactsSyncing.m (hunk 2 above)
Comment (charlesmchen) [inline Signal/src/util/OWSContactsSyncing.m]: Done.

File: Signal/src/util/OWSContactsSyncing.m (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/util/OWSContactsSyncing.m]: Done.

**Pull request merged**